community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NumberKey.java

```diff
@@ -0,0 +1,200 @@
+        this.entityId = entityId;
+        entityIdIsSpecialTieBreaker = false;
+    }
+
+    String propertiesAsString()
+    {
+        return String.valueOf( toNumberValue() );
+    }
+
+    void initAsLowest()
+    {
+        rawValueBits = Double.doubleToLongBits( Double.NEGATIVE_INFINITY );
+        type = TYPE_DOUBLE;
+        entityId = Long.MIN_VALUE;
+        entityIdIsSpecialTieBreaker = true;
+    }
+
+    void initAsHighest()
+    {
+        rawValueBits = Double.doubleToLongBits( Double.POSITIVE_INFINITY );
+        type = TYPE_DOUBLE;
+        entityId = Long.MAX_VALUE;
+        entityIdIsSpecialTieBreaker = true;
+    }
+
+    /**
+     * Compares the value of this key to that of another key.
+     * This method is expected to be called in scenarios where inconsistent reads may happen (and later retried).
+     *
+     * @param other the {@link NumberKey} to compare to.
+     * @return comparison against the {@code other} {@link NumberKey}.
+     */
+    int compareValueTo( NumberKey other )
+    {
+        return type == TYPE_LONG && other.type == TYPE_LONG
+                // If both are long values then compare them directly, w/o going through double.
+                // This is because at high values longs have higher precision, or double lower rather,
+                // than double values, so converting them to doubles and comparing would have false positives.
+                ? Long.compare( rawValueBits, other.rawValueBits )
+
+                // Otherwise convert both to double and compare, with the reasoning that the long precision
+                // cannot be upheld anyway and double precious being higher than float precision.
+                : Double.compare( doubleValue(), other.doubleValue() );
+    }
+
+    /**
+     * @return the value as double, with potential precision loss.
+     */
+    private double doubleValue()
+    {
+        switch ( type )
+        {
+        case TYPE_LONG:
+            return rawValueBits;
+        case TYPE_FLOAT:
+            return Float.intBitsToFloat( (int) rawValueBits );
+        case TYPE_DOUBLE:
+            return Double.longBitsToDouble( rawValueBits );
+        default:
+            // This is interesting: because of the nature of the page cache and the point in time this method
+            // is called we cannot really throw exception here if type is something unexpected - it may simply
+            // have been an inconsistent read, which will be retried.
+            // It's not for us to decide here, so let's return NaN here.
+            return Double.NaN;
+        }
+    }
+
+    /**
+     * Extracts data from a {@link Number} into state of this {@link NumberKey} instance.
+     *
+     * @param value actual {@link Number} value.
+     */
+    private void extractValue( Number value )
+    {
+        if ( value instanceof Double )
+        {
+            type = TYPE_DOUBLE;
+            rawValueBits = Double.doubleToLongBits( (Double) value );
```

Comment:
This collapses all NaN values to a single canonical NaN value, meaning all NaNs are the same. I'm not sure what we are supposed to promise in this situation.